File: 0079_wordSearch/main.cpp
#include <vector>
#include <string>

using namespace std;

class Solution {
public:
    bool exist(vector<vector<char>>& board, string word) {
        for(int i = 0; i < board.size(); ++i) {
            for(int j = 0; j < board[i].size(); ++j) {
                if(helper(board, word, 0, i, j))
                    return true;
            }
        }
        return false;
    }

    bool helper(vector<vector<char>>& board, string word, int currentIdx, int currentRow, int currentCol) {
        // match all char
        if(currentIdx == word.size())
            return true;

        // out of bound
        if(currentRow < 0 || currentRow >= board.size() || currentCol < 0 || currentCol >= board[currentRow].size())
            return false;

        // current char not match the board
        if(word[currentIdx] != board[currentRow][currentCol])
            return false;
        
        char temp = board[currentRow][currentCol];
        board[currentRow][currentCol] = NULL;

        bool res = helper(board, word, currentIdx+1, currentRow+1, currentCol) || helper(board, word, currentIdx+1, currentRow-1, currentCol) ||
                    helper(board, word, currentIdx+1, currentRow, currentCol+1) || helper(board, word, currentIdx+1, currentRow, currentCol-1);
        
        board[currentRow][currentCol] = temp;

        return res;
    }
};

File: 0079_wordSearch/main.py
from typing import List

class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        visited = [[False for _ in range(len(board[0]))] for _ in range(len(board))]
        def dfs(r, c, ith):
            if ith == len(word):
                return True
            if r < 0 or r >= len(board) or c < 0 or c >= len(board[0]) or board[r][c] != word[ith] or visited[r][c]:
                return False
                
            visited[r][c] = True
            top = dfs(r - 1, c, ith + 1)
            down = dfs(r + 1, c, ith + 1)
            left = dfs(r, c - 1, ith + 1)
            right = dfs(r, c + 1, ith + 1)
            visited[r][c] = False

            return top or down or left or right
        
        for r in range(len(board)):
            for c in range(len(board[0])):
                if dfs(r, c, 0):
                    return True
        return False